Move HSFLL configuration to DT

gmarull opened the pull request:
Note: a simple clock control driver (init only?) could also be implemented to avoid code in soc.

gmarull commented:
cc: @57300

57300 commented on dts/common/nordic/nrf54h20_eng0.dtsi:
```diff
@@ -63,6 +63,12 @@
+			compatible = "nordic,nrf-ficr";
+			reg = <0xfffe000 0x800>;
+			#ficr-cells = <1>;
+		};
```
nit: Maybe move under `cpurad_uicr`, to keep the nodes sorted by unit address. Could use `DT_SIZE_K(2)` as well.

gmarull replied:
done

57300 commented on soc/arm/nordic_nrf/nrf54h/soc.c:
```diff
@@ -15,14 +15,24 @@ LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
+   DT_PHA_BY_NAME(HSFLL_NODE, ficrs, coarse, offset)
+#define FICR_HSFLL_FINE                                                        \
+   DT_REG_ADDR(DT_PHANDLE_BY_NAME(HSFLL_NODE, ficrs, fine)) +                  \
+   DT_PHA_BY_NAME(HSFLL_NODE, ficrs, fine, offset)
```
Will need clang-format.

gmarull replied:
should be better now

anangl commented on include/zephyr/dt-bindings/misc/nordic-nrf-ficr-nrf54h20-eng0.h:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
```
2024

gmarull commented on soc/arm/nordic_nrf/nrf54h/soc.c:
```diff
@@ -53,23 +57,21 @@ static void power_domain_init(void)
 	LOG_DBG("Trim: HSFLL VSUP: 0x%.8x", trim.vsup);
 	LOG_DBG("Trim: HSFLL COARSE: 0x%.8x", trim.coarse);
 	LOG_DBG("Trim: HSFLL FINE: 0x%.8x", trim.fine);
 
-	if (CONFIG_HSFLL_CLOCK_MULTIPLIER) {
-		nrf_hsfll_clkctrl_mult_set(hsfll, CONFIG_HSFLL_CLOCK_MULTIPLIER);
-	}
+	nrf_hsfll_clkctrl_mult_set(hsfll, DT_PROP(HSFLL_NODE, multiplier));
```
note here: multiplier is only applicable when working in closed loop mode, however, I assume configuring it to the right value won't hurt. This code snippet doesn't seem to configure the operating mode.

gmarull commented on dts/bindings/clock/nordic,nrf-hsfll.yaml:
```diff
@@ -0,0 +1,71 @@
+    description: |
+      Clock multiplication factor. Defaults to 6, reset default.
+
+  trim:
+    type: boolean
+    description: |
+      Enable trimming of the HSFLL VSUP, COARSE and FINE parameters.
```
this is really a software config, wondering if Kconfig should be kept for it instead of using a DT property, thoughts?

anangl replied:
How about dropping this property and renaming the two others to `trims` and `trim-names`, and performing the trimming based on the presence of the `trims` property?

gmarull replied:
updated to kconfig

anangl replied:
> How about dropping this property and renaming the two others to `trims` and `trim-names`

That won't work, right? The property specifying FICR entries must be named `ficrs`.

gmarull replied:
you can make use of `specifier-space` bindings property, but it looks like `names` doesn't work then...

gmarull replied:
The problem I see with using the presence of ficrs/trims is that these are defined in soc dts files for convenience (so always there). I wonder is there is any scenario where we don't want to trim though... :-)

anangl replied:
> The problem I see with using the presence of ficrs/trims is that these are defined in soc dts files for convenience (so always there).

One could delete this property in an overlay if really needed.

> I wonder is there is any scenario where we don't want to trim though... :-)

Exactly. :-)

anangl commented on soc/arm/nordic_nrf/nrf54h/soc.c:
```diff
@@ -53,23 +57,21 @@ static void power_domain_init(void)
-		.coarse = FICR_TRIM_BLOCK.HSFLL.TRIM.COARSE[CONFIG_FICR_HSFLL_TRIM_IDX],
-		.fine = FICR_TRIM_BLOCK.HSFLL.TRIM.FINE[CONFIG_FICR_HSFLL_TRIM_IDX],
-	};
+	nrf_hsfll_trim_t trim;
+
+	trim.vsup = sys_read32(FICR_ADDR_GET(HSFLL_NODE, vsup));
```
I'd keep the full initialization of the structure, as it's generally safer (I don't expect this particular structure to be expanded in the future, but who knows).
```suggestion
	nrf_hsfll_trim_t trim = {
		.vsup = sys_read32(FICR_ADDR_GET(HSFLL_NODE, vsup)),
```

anangl commented on dts/common/nordic/nrf54h20_eng0.dtsi:
```diff
@@ -77,6 +79,12 @@
 			ptr-ext-uicr = <&cpurad_uicr_ext>;
 		};
 
+		ficr: memory@fffe000 {
```
Since we use `uicr` for UICRs, wouldn't it be better to use `ficr` here? It's a specific kind of memory. Now I see that also in old SoCs the nodes are named `ficr`, so let's be consistent.

gmarull replied:
I think `memory` fits better (following dt spec and its usage), but using ficr is fine I guess

![image](https://github.com/anangl/zephyr/assets/25011557/78f4d283-0324-4d76-b23c-95d086f9e45e)

57300 replied:
Technically, *ICRs are just memories, but I think our documentation usually describes them as peripherals.

gmarull commented on dts/bindings/clock/nordic,nrf-hsfll.yaml:
```diff
@@ -0,0 +1,66 @@
+    description: |
+      Names of the FICR entries used to trim the HSFLL VSUP, COARSE and FINE
+      parameters. It must contain "vsup", "coarse" and "fine". Required if trim
+      is enabled.
```
note: in the future these could be moved to a common file, re-usable by other bindings (e.g. CAN PLL will need this as well).

57300 commented on soc/arm/nordic_nrf/nrf54h/Kconfig.series:
```diff
@@ -16,30 +16,10 @@ config SOC_SERIES_NRF54HX
-	int
-	depends on SOC_TRIM_HSFLL
+config NRF_HSFLL_TRIM
+	bool "Trim HSFLL clock on startup"
```
Not sure if this should have a prompt, at least in public-facing code.

gmarull replied:
Then we may just drop this entirely and always trim :-) Who is the right person to ask if trimming is really optional?

kyberdin replied:
HSFLL trimming is a hard requirement on actual hardware, and the domains themselves are responsible for setting that up.

gmarull replied:
so should that option be dropped?

gmarull replied:
added a new commit that drops this option.

gmarull commented:
last push: dropped multiplier in favor of clock-frequency property, I think it's more descriptive. Also added FLL16M so we have a more accurate clock tree (even if FLL16M is entirely managed by sysctrl).

anangl commented on dts/bindings/misc/nordic-nrf-ficr-client.yaml:
```diff
@@ -0,0 +1,14 @@
+    type: phandle-array
+    description: |
+      FICR entries, e.g. <&ficr OFFSET>. Available offsets (or FICR entries) are
+      available at <zephyr/dt-bingins/misc/nordic-nrf-ficr-*.h>.
```
```suggestion
      available at <zephyr/dt-bindings/misc/nordic-nrf-ficr-*.h>.
```

anangl commented on dts/bindings/clock/nordic,nrf-hsfll.yaml:
```diff
@@ -0,0 +1,65 @@
+      - 384000000
+      - 400000000
+
+  nordic,ficrs:
+    required: true
+
+  nordic,ficr-names:
+    required: true
```
Couldn't these properties be set as required in `nordic-nrf-ficr-client.yaml`?

gmarull replied:
hmm what if some clients want them optional (e.g. conditioned to another property being set)?

anangl replied:
I just thought that it's a bit strange if a FICR client does not define at least the `nordic,ficrs` property. But let's keep it as it is for now.

kyberdin replied:
> hmm what if some clients want them optional (e.g. conditioned to another property being set)?

Can you clarify on what you mean here a bit more? Is it optional in the sense of just being linked to another property, rather than user-defined?

gmarull replied:
So we could have a binding that only required FICR entries under certain conditions. For example, imagine we have a PLL with a `nordic,mode` property. If `mode` is "open", FICR is not read by the driver, whereas if "closed", some FICR tune values are needed. I'm not sure if a case like that will ever happen, but I assume it's not impossible that's why I haven't made the requirement at client YAML file. It's relatively similar to `pinctrl-device.yaml`.

gmarull replied:
Note that our binding language doesn't allow us to do fancy things like "required if property X is present", and added to the fact that required cannot be undone, we must play safe here.

kyberdin replied:
AFAICT from spec, FICR values are always going to be needed by these drivers for retrieving the production-defined default values to load into the HSFLL.TRIM registers, so I'm not sure if the case would happen either. But I understand the need for flexibility given the language constraints though, and I'm all for leaving that option open for us in the future so doing that here now makes sense.

My only suggestion then would be that if we do leave FICR things open as "optional", to maybe include some clarification in a binding's description text they're not optional from a user configuration standpoint (because they aren't pick-and-choose settings and are there for hardware operation), just for driver flexibility.

57300 commented:
> Also added FLL16M so we have a more accurate clock tree (even if FLL16M is entirely managed by sysctrl).

I don't think FLL16M can be exposed. Also, adding the node now will confuse nrf-regtool.

kyberdin commented on include/zephyr/dt-bindings/misc/nordic-nrf-ficr-nrf54h20-eng0.h:
```diff
@@ -0,0 +1,37 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_MISC_NORDIC_NRF_FICR_NRF54H20_ENG0_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_MISC_NORDIC_NRF_FICR_NRF54H20_ENG0_H_
+
+#define NRF_FICR_TRIM_APPLICATION_HSFLL_TRIM_VSUP     0x3D8U
```
I strongly advise against hardcoding register offsets that are delivered through the MDK, especially when it regards device-critical contents (see comment below).

gmarull replied:
MDK cannot be pulled into dt-bindings/, so it will need to be done this way. Note that other OSes like Linux have worked for decades without MDK, SVD etc. and have managed to support far more complex systems, arguably better than Zephyr.

kyberdin replied:
> MDK cannot be pulled into dt-bindings/, so it will need to be done this way.

I realize we can't include MDK headers this way, nor was I suggesting that.

> Note that other OSes like Linux have worked for decades without MDK, SVD etc. and have managed to support far more complex systems, arguably better than Zephyr.

Apologies for not being more clear, I'm not trying to say otherwise. For Nordic, redefining register offsets in separate files opens the door to misalignment against the MDK that its HAL aligns with. My comment here advises against this.

masz-nordic replied:
I agree with @kyberdin - using hardcoded values in DT is not future-proof and should be avoided.

gmarull replied:
FYI https://github.com/zephyrproject-rtos/hal_nordic/pull/161, so I guess we can close this topic

kyberdin commented on soc/arm/nordic_nrf/nrf54h/soc.c:
```diff
@@ -53,23 +58,22 @@ static void power_domain_init(void)
-		.vsup = FICR_TRIM_BLOCK.HSFLL.TRIM.VSUP,
-		.coarse = FICR_TRIM_BLOCK.HSFLL.TRIM.COARSE[CONFIG_FICR_HSFLL_TRIM_IDX],
-		.fine = FICR_TRIM_BLOCK.HSFLL.TRIM.FINE[CONFIG_FICR_HSFLL_TRIM_IDX],
+	nrf_hsfll_trim_t trim = {
+		.vsup = sys_read32(FICR_ADDR_GET(HSFLL_NODE, vsup)),
+		.coarse = sys_read32(FICR_ADDR_GET(HSFLL_NODE, coarse)),
+		.fine = sys_read32(FICR_ADDR_GET(HSFLL_NODE, fine))
```
This is quite a sleek way of accessing FICR things in a general manner, but IMO it is too nice and generic for how FICR should be accessed.

By maintaining the register offset in a devicetree include, we've now abstracted where that value is accessed to outside of the HAL, and must ensure by hand that the defined offsets align with what is given in the MDK. Otherwise we risk loading invalid contents into a place that can have a major impact on device performance. I cannot stress enough how much we should avoid trim-related issues.

For the 54H SOC, the only required information for setting these are {VSUP, FINE, COARSE}, and the index of each that is applicable to the HSFLL. Since this is SOC-specific, the existing `FICR_TRIM_BLOCK.HSFLL.TRIM.{type}[index]` approach won't step on other SOCs' toes and is the safer one to ensure that the register is always accessed at the location defined by spec.

Perhaps it would be possible to have the ficr cell be a list of `[type-enum, index]` instead of `offset`? If the type is enumerated, we can have it contain many different types (depending on how FICR usage and the need of its bindings grow) that can then be translated into use cases within the `soc.c` code.

gmarull replied:
AFAIK DT cells are just integer values that have a particular name (offset in this case), so we cannot enumerate. Also, `index` would only be useful for a few entries, I see most are just scalar entries.

gmarull commented:
> > Also added FLL16M so we have a more accurate clock tree (even if FLL16M is entirely managed by sysctrl).
> 
> I don't think FLL16M can be exposed. Also, adding the node now will confuse nrf-regtool.

HSFLL documents that is fed with FLL16M, so I guess it's publicly known that we have a 16 MHz DCO. What is not exposed is control of it, something this PR doesn't attempt to do. I see no better way to express the clock tree than adding all clocks in DT, the alternative is hardcoding 16000000UL in a C file, which in the end also exposes HSFLL is fed with a 16 MHz clock...

57300 commented:
> What is not exposed is control of it

You do expose its address. Can it be removed?

The clock itself should probably move out of `global_peripherals` as well. Maybe the standard `/clocks` node would work, although I can't find much info about it.

57300 commented:
Just a thought: should HSFLL configuration relate to the `clock-frequency` property of `/cpus/cpu@x` nodes somehow?

gmarull commented:
> > What is not exposed is control of it
> 
> You do expose its address. Can it be removed?

It can be done, but I guess sysctrl won't be able to re-use these dts files (now it probably doesn't even use DT for FLL16M address, but it should likely be done).

> 
> The clock itself should probably move out of `global_peripherals` as well. Maybe the standard `/clocks` node would work, although I can't find much info about it.

gmarull commented:
> Just a thought: should HSFLL configuration relate to the `clock-frequency` property of `/cpus/cpu@x` nodes somehow?

I don't know, clocks in Zephyr are a bit of a mess. Some CPU nodes define the CPU frequency, which is basically put into a Kconfig option, not really used for anything else. Here it is used to compute the right multiplier in soc code manipulating HSFLL.

57300 commented:
> I guess sysctrl won't be able to re-use these dts files

Sysctrl will include an additional dtsi with private contents.

gmarull commented:
> > I guess sysctrl won't be able to re-use these dts files
> 
> Sysctrl will include an additional dtsi with private contents.

But if we define a node with nodelabel fll16m, it will conflict if they want to re-use that, right? I guess we need a criteria to define the _public_ facing clocks (ie, a fixed-clock, non-configurable), and _private_ clocks (the configurable one). I guess it makes sense to define FLL16M as a fixed-clock in clocks node, but we should probably _namespace_ them (or sysctrl namespace its private entries)

57300 commented:
> > > I guess sysctrl won't be able to re-use these dts files
> > 
> > 
> > Sysctrl will include an additional dtsi with private contents.
> 
> But if we define a node with nodelabel fll16m, it will conflict if they want to re-use that, right? I guess we need a criteria to define the _public_ facing clocks (ie, a fixed-clock, non-configurable), and _private_ clocks (the configurable one). I guess it makes sense to define FLL16M as a fixed-clock in clocks node, but we should probably _namespace_ them (or sysctrl namespace its private entries)

Hmm, I was actually thinking of just overlaying the `reg` property and ignoring the `unit_address_vs_reg` warning, but Zephyr promotes it to an error for some reason. Anyway, let's evaluate this internally.

masz-nordic reviewed (changes requested):
Hardcoded values taken from PS should not be used, some solution using MDK structures is needed here.

gmarull commented:
> Hardcoded values taken from PS should not be used, some solution using MDK structures is needed here.

Scripts to autogenerate the dt-bindings file are welcome.

masz-nordic commented:
>Scripts to autogenerate the dt-bindings file are welcome.

Indeed. But as they do not exist yet, we should stick to what we have, that is MDK structures.

gmarull commented:
> > Scripts to autogenerate the dt-bindings file are welcome.
> 
> Indeed. But as they do not exist yet, we should stick to what we have, that is MDK structures.

Again, MDK structures can't be pulled into `dt-bindings`. This shouldn't prevent doing things properly from a DT perspective.

gmarull commented:
For reference, other vendors that use a strategy similar to FICR access: https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/power/ambiq%2Cpwrctrl.yaml. In this case, the vendor doesn't even provide helper definitions and uses hardcoded values in dtsi files (see dts/arm/ambiq/ambiq_apollo4p_blue.dtsi).

masz-nordic commented:
>This shouldn't prevent doing things properly from a DT perspective.

This was discussed many times, and the decision was to use MDK until the devicetree is generated from PS.

gmarull commented:
> > This shouldn't prevent doing things properly from a DT perspective.
> 
> This was discussed many times, and the decision was to use MDK until the devicetree is generated from PS.

Other vendors don't get such exceptions upstream, and so we won't either (ambiq for example, was forced to do that despite proposing an inferior solution using their HAL).

gmarull commented:
last push: ficr offsets autogenerated from SVD, FLL16M is now a `fixed-clock` (no longer a memory-mapped device)

57300 commented on soc/arm/nordic_nrf/nrf54h/soc.c:
```diff
@@ -15,14 +16,18 @@ LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
 #error "Unsupported SoC"
 #endif
 
+#define HSFLL_TRIM DT_PROP(HSFLL_NODE, trim)
```
Leftover.

gmarull replied:
thanks, fixed

57300 reviewed (approved):
LGTM. Some changes can be squashed, at least the Kconfig ones.

gmarull commented:
> LGTM. Some changes can be squashed, at least the Kconfig ones.

done. @anangl should be ready to integrate

anangl commented:
> > LGTM. Some changes can be squashed, at least the Kconfig ones.
> 
> done. @anangl should be ready to integrate

@gmarull I've just updated my PR. I squashed some of the commits into existing ones. I hope that's okay.

gmarull commented:
closing as it's now integrated
